File: backend/python-flask/app/models/itinerary.py
import uuid
from datetime import datetime
from app.db import db


class Itinerary(db.Model):
    __tablename__ = 'itineraries'

    id = db.Column(db.String(50), primary_key=True, default=lambda: str(uuid.uuid4()))
    created_at = db.Column(db.String(30), nullable=False, default=lambda: datetime.utcnow().isoformat() + 'Z')
    strategy = db.Column(db.String(50), nullable=False)
    total_distance = db.Column(db.Float, nullable=False)

    stops = db.relationship('ItineraryStop', backref='itinerary', order_by='ItineraryStop.stop_number')

    def to_dict(self):
        return {
            'id': self.id,
            'createdAt': self.created_at,
            'strategy': self.strategy,
            'totalDistance': self.total_distance,
            'stops': [stop.to_dict() for stop in self.stops],
        }


class ItineraryStop(db.Model):
    __tablename__ = 'itinerary_stops'

    id = db.Column(db.String(50), primary_key=True, default=lambda: str(uuid.uuid4()))
    itinerary_id = db.Column(db.String(50), db.ForeignKey('itineraries.id'), nullable=False)
    stop_number = db.Column(db.Integer, nullable=False)
    match_id = db.Column(db.String(50), db.ForeignKey('matches.id'), nullable=False)
    city_id = db.Column(db.String(50), db.ForeignKey('cities.id'), nullable=False)
    distance_from_previous = db.Column(db.Float, nullable=False, default=0)

    match = db.relationship('Match', foreign_keys=[match_id])
    city = db.relationship('City', foreign_keys=[city_id])

    def to_dict(self):
        return {
            'stopNumber': self.stop_number,
            'match': self.match.to_dict() if self.match else None,
            'city': self.city.to_dict() if self.city else None,
            'distanceFromPrevious': self.distance_from_previous,
        }

File: backend/node-express/src/models/Itinerary.ts
import db from '../db/connection';
import { v4 as uuidv4 } from 'uuid';
import { OptimisedRoute } from '../strategies/RouteStrategy';

interface ItineraryRow {
  id: string;
  created_at: string;
  strategy: string;
  total_distance: number;
}

interface StopRow {
  id: string;
  itinerary_id: string;
  stop_number: number;
  match_id: string;
  city_id: string;
  distance_from_previous: number;
}

export function create(route: OptimisedRoute): { id: string } {
  const id = uuidv4();
  const createdAt = new Date().toISOString();

  const insertItinerary = db.prepare(
    'INSERT INTO itineraries (id, created_at, strategy, total_distance) VALUES (?, ?, ?, ?)'
  );
  insertItinerary.run(id, createdAt, route.strategy, route.totalDistance);

  const insertStop = db.prepare(
    'INSERT INTO itinerary_stops (id, itinerary_id, stop_number, match_id, city_id, distance_from_previous) VALUES (?, ?, ?, ?, ?, ?)'
  );

  for (const stop of route.stops) {
    insertStop.run(
      uuidv4(),
      id,
      stop.stopNumber,
      stop.match.id,
      stop.city.id,
      stop.distanceFromPrevious
    );
  }

  return { id };
}

export function getById(id: string): (ItineraryRow & { stops: StopRow[] }) | undefined {
  const itinerary = db.prepare('SELECT * FROM itineraries WHERE id = ?').get(id) as
    | ItineraryRow
    | undefined;

  if (!itinerary) return undefined;

  const stops = db
    .prepare(
      'SELECT * FROM itinerary_stops WHERE itinerary_id = ? ORDER BY stop_number'
    )
    .all(id) as StopRow[];

  return { ...itinerary, stops };
}
